### PR TITLE
Switch placement to statefulset for multiple placement replicas - WIP

### PR DESCRIPTION
**Pull request by youngbupark**
# Description

Switch placement to statefulset for multiple placement replicas

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2376 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by youngbupark** [comment]
@yaron2 @artursouza @mukundansundar @tcnghia I am running the scale test for new changes. 

leader election for placement service is working as expected. Some other fixes will come together while running the test.

I am trying to scale actor service instance to 1000.

**Comment by yaron2** [comment]
> @yaron2 @artursouza @mukundansundar @tcnghia I am running the scale test for new changes.
> 
> leader election for placement service is working as expected. Some other fixes will come together while running the test.
> 
> I am trying to scale actor service instance to 1000.

We should probably document the guidelines for how to deploy actor services. The general rule is that instead of a large number of small pods, you'd want a fewer number of pods with more CPU/Memory requests.

Currently, 1000 is an arbitrary number.
I highly suggest that *before* trying to scale to X number of pods, we instead opt for a number of actors in a cluster, which is a metric that's more usable.

In essence, this means finding out how many actors can Dapr activate in a single pod given X vCPUs and Y memory, and then deciding on testing for the number of pods needed to run an incrementally growing number of actors: 10k, 100k, 500k, 1M, 10M.

This makes more sense than trying to scale to 1000 pods, IMO.

File: charts/dapr/charts/dapr_placement/templates/dapr_placement_deployment.yaml [hunk 1]
```diff
@@ -108,3 +137,14 @@ spec:
...
+{{- if eq .Values.global.ha.enabled true }}
+  volumeClaimTemplates:
+  - metadata:
+      name: raft-log
+    spec:
+      accessModes:
+        - ReadWriteOnce
+      resources:
+        requests:
+          storage: 1Gi
```
Comment (mukundansundar) [inline charts/dapr/charts/dapr_placement/templates/dapr_placement_deployment.yaml:149]: shouldn't this be a parameter?

File: pkg/injector/pod_patch.go [hunk 2]
```diff
@@ -114,7 +114,7 @@ func (i *injector) getPodPatchOperations(ar *v1.AdmissionReview,
 	}
 
 	// Keep DNS resolution outside of getSidecarContainer for unit testing.
-	placementAddress := fmt.Sprintf("%s:80", getKubernetesDNS(placementService, namespace))
+	placementAddress := fmt.Sprintf("%s:50005", getKubernetesDNS(placementService, namespace))
```
Comment (mukundansundar) [inline pkg/injector/pod_patch.go:117]: will this resolve to multiple addresses?

File: pkg/placement/raft/server.go [hunk 3]
```diff
@@ -69,6 +69,23 @@ func New(id string, inMem bool, peers []PeerInfo, logStorePath string) *Server {
...
 
+func tryResolveRaftAdvertiseAddr(bindAddr string) (*net.TCPAddr, error) {
+	// HACKHACK: Kubernetes POD DNS A record population takes some time
+	// to look up the address after StatefulSet POD is deployed.
+	const retryCount = 120
+	var interval = 1 * time.Second
+
+	var err error
+	for retry := 0; retry < retryCount; retry++ {
+		addr, err := net.ResolveTCPAddr("tcp", bindAddr)
```
Comment (mukundansundar) [inline pkg/placement/raft/server.go:80]: add back todo for `tls`

**Comment by youngbupark** [comment]
> > @yaron2 @artursouza @mukundansundar @tcnghia I am running the scale test for new changes.
> > leader election for placement service is working as expected. Some other fixes will come together while running the test.
> > I am trying to scale actor service instance to 1000.
> 
> We should probably document the guidelines for how to deploy actor services. The general rule is that instead of a large number of small pods, you'd want a fewer number of pods with more CPU/Memory requests.
> 
> Currently, 1000 is an arbitrary number.

I think we talked about this number in the previous spec review. The current focus is the reliability of placement and runtime. e.g. no crashes and no mem leak.

> I highly suggest that _before_ trying to scale to X number of pods, we instead opt for a number of actors in a cluster, which is a metric that's more usable.
> 

The perf and actual tests regarding below would be next work item once we have reliability of placement and runtime. 

> In essence, this means finding out how many actors can Dapr activate in a single pod given X vCPUs and Y memory, and then deciding on testing for the number of pods needed to run an incrementally growing number of actors: 10k, 100k, 500k, 1M, 10M.
> 
> This makes more sense than trying to scale to 1000 pods, IMO.

File: charts/dapr/charts/dapr_placement/templates/dapr_placement_deployment.yaml (hunk 1 above)
Comment (youngbupark) [reply charts/dapr/charts/dapr_placement/templates/dapr_placement_deployment.yaml:149]: yup this is WIP.

File: pkg/injector/pod_patch.go (hunk 2 above)
Comment (youngbupark) [reply pkg/injector/pod_patch.go:117]: right. please read kubernetes headless service. we use this way for service invocation.

**Comment by yaron2** [comment]
> > > @yaron2 @artursouza @mukundansundar @tcnghia I am running the scale test for new changes.
> > > leader election for placement service is working as expected. Some other fixes will come together while running the test.
> > > I am trying to scale actor service instance to 1000.
> > 
> > 
> > We should probably document the guidelines for how to deploy actor services. The general rule is that instead of a large number of small pods, you'd want a fewer number of pods with more CPU/Memory requests.
> > Currently, 1000 is an arbitrary number.
> 
> I think we talked about this number in the previous spec review. The current focus is the reliability of placement and runtime. e.g. no crashes and no mem leak.
> 
> > I highly suggest that _before_ trying to scale to X number of pods, we instead opt for a number of actors in a cluster, which is a metric that's more usable.
> 
> The perf and actual tests regarding the below would be next work item once we have reliability of placement and runtime.
> 
> > In essence, this means finding out how many actors can Dapr activate in a single pod given X vCPUs and Y memory, and then deciding on testing for the number of pods needed to run an incrementally growing number of actors: 10k, 100k, 500k, 1M, 10M.
> > This makes more sense than trying to scale to 1000 pods, IMO.

> I think we talked about this number in the previous spec review. The current focus is the reliability of placement and runtime. e.g. no crashes and no mem leak.

Right, but again: why 1000 and not 100? or 100000? what I'm saying is, you need to make sure you test against a realistic number of pods. I'm saying we don't know what realistic is at this point.

**Comment by youngbupark** [comment]
> 
> Right, but again: why 1000 and not 100? or 100000? what I'm saying is, you need to make sure you test against a realistic number of pods. I'm saying we don't know what realistic is at this point.

I am not just deploying 1000 pods. I am starting from 1 pod to 1000 pods by increasing 10 - 50 pods gradually. I am trying to see how placement and runtime work in terms of resources usage and reliability. I know 1000 PODs would not be realistic. It could hit cluster's limit. I think in the previous meeting, you argued about this number, but I said *trying* to scale out to 1000 PODs. The key in the validation is to see how placement handles these amount of PODs. not just to scale to 1000 PODs as you know.

**Comment by youngbupark** [comment]
While testing, I fix some bugs and improve the memory usage and performance. I will split this PR from helm chart change.

**Pull request closed without merging**